fabric-client-gametest-api-v1/src/client/java/net/fabricmc/fabric/api/client/gametest/v1/TestScreenshotComparisonAlgorithm.java

```diff
@@ -0,0 +1,128 @@
+	}
+
+	/**
+	 * The Mean Squared Difference algorithm computes the mean of differences between the needle pixels and
+	 * corresponding haystack pixels squared, and checks if the mean is less than a particular threshold.
+	 *
+	 * @param maxMeanSquaredDifference The maximum mean squared difference between the needle and the subimage in the
+	 *                                 haystack for a match
+	 * @return The Mean Squared Difference algorithm with the given threshold
+	 */
+	static TestScreenshotComparisonAlgorithm meanSquaredDifference(float maxMeanSquaredDifference) {
+		Preconditions.checkArgument(maxMeanSquaredDifference >= 0 && maxMeanSquaredDifference <= 1, "maxMeanSquaredError must be between 0 and 1");
+		return new TestScreenshotComparisonAlgorithms.MeanSquaredDifference(maxMeanSquaredDifference);
+	}
+
+	/**
+	 * An algorithm which searches for an exact match of the needle in the haystack. This is simpler and faster than a
+	 * fuzzy match, but is prone to failing on even the slightest inconsistency in the image.
+	 *
+	 * @return An algorithm for exact matches
+	 */
+	static TestScreenshotComparisonAlgorithm exact() {
+		return TestScreenshotComparisonAlgorithms.Exact.INSTANCE;
+	}
+
+	/**
+	 * Finds a needle in a haystack, color image version. Pixels are in RGB format (with no alpha channel).
+	 *
+	 * @param haystack The image in which to search for the needle
+	 * @param needle The image to search for
+	 * @return The location of the match, or {@code null} if no match was found. In case of multiple matches, returns an
+	 *         arbitrary match
+	 */
+	@Nullable
+	Vector2i findColor(RawImage<int[]> haystack, RawImage<int[]> needle);
+
+	/**
+	 * Finds a needle in a haystack, grayscale image version. Each pixel is a brightness value from 0-255.
+	 *
+	 * <p>Note for custom algorithm implementations: the default implementation converts the images to color and then
+	 * uses {@link #findColor}, but it is usually possible to implement this more efficiently. You should strongly
+	 * consider overriding this method.
+	 *
+	 * @param haystack The image in which to search for the needle
+	 * @param needle The image to search for
+	 * @return The location of the match, or {@code null} if no match was found. In case of multiple matches, returns an
+	 *         arbitrary match
+	 */
+	@Nullable
+	default Vector2i findGrayscale(RawImage<byte[]> haystack, RawImage<byte[]> needle) {
+		RawImage<int[]> colorHaystack = TestScreenshotComparisonAlgorithms.RawImageImpl.toColor(haystack);
+		RawImage<int[]> colorNeedle = TestScreenshotComparisonAlgorithms.RawImageImpl.toColor(needle);
+		return findColor(colorHaystack, colorNeedle);
+	}
+
+	/**
+	 * A thin wrapper around a raw image data array, so that algorithms can directly access the array which may be more
+	 * efficient than going through a {@link NativeImage} each time.
+	 *
+	 * @param <DATA> The type of the image data array
+	 */
+	@ApiStatus.NonExtendable
+	interface RawImage<DATA> {
```

Comment:
This is a non-extendable interface, but are there any public APIs that create instances of it (ie can it even be used with API code)?

Reply:
There are no public APIs to create instances of `RawImage`, but it can be used by custom implementations of `TestScreenshotComparisonAlgorithm`